File: src/_03_BehavioralPattern/_03_11_Visitor/Phone.js
import Device from "./Device.js";

import Rectangle from "./Rectangle.js";
import Triangle from "./Triangle.js";
import Circle from "./Circle.js";
import Star from "./Star.js";

/**
 * @typedef { import("./Shape").default } Shape
 */

export default class Phone extends Device {
  constructor() {
    super();
  }
  
  /**
   * @override
   * @param { Shape } shape 
   */
  print(shape) {
    switch(shape.constructor) {
      case Rectangle: {
        console.log("- Print Rectangle To Phone");
        break;
      }

      case Triangle: {
        console.log("- Print Triangle To Phone");
        break;
      }

      case Circle: {
        console.log("- Print Circle To Phone");
        break;
      }

      case Star: {
        console.log("- Print Star To Phone");
        break;
      }
    }
  }
}

File: src/_03_BehavioralPattern/_03_11_Visitor/java/after/Phone.java
package _03_BehavioralPattern._03_11_Visitor.java.after;

public class Phone implements Device {

  @Override
  public void print(Rectangle rectangle) {
    System.out.println("- Print Rectangle to Phone");
  }

  @Override
  public void print(Circle circle) {
    System.out.println("- Print Circle To Phone");
  }

  @Override
  public void print(Triangle triangle) {
    System.out.println("- Print Triangle To Phone");
  }

  @Override
  public void print(Star start) {
    System.out.println("- Print Star To Phone");
  }
  
}
